File: formacOS/Tests/AnyClipDaemonTests/InteropTests.swift
import Testing
import Foundation
import Network
@testable import AnyClipDaemon
@testable import AnyClipCore

private func scriptsDir() -> URL {
    // <pkg>/Tests/AnyClipDaemonTests/InteropTests.swift -> <pkg>/Scripts
    URL(fileURLWithPath: #filePath)
        .deletingLastPathComponent()   // AnyClipDaemonTests
        .deletingLastPathComponent()   // Tests
        .deletingLastPathComponent()   // formacOS
        .appendingPathComponent("Scripts")
}

@Test func interopWithPythonFakePeer() async throws {
    let port: UInt16 = 28491
    let outFile = FileManager.default.temporaryDirectory
        .appendingPathComponent("fake-peer-\(UUID().uuidString).jsonl")

    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = [
        "python3", scriptsDir().appendingPathComponent("fake_peer.py").path,
        "--port", "\(port)", "--token", "interop-token",
        "--out", outFile.path,
    ]
    let stdout = Pipe()
    process.standardOutput = stdout
    try process.run()
    defer { if process.isRunning { process.terminate() } }

    // Wait for READY with a deadline to avoid flakiness from availableData
    // returning before READY is written.
    var readyReceived = false
    let readyDeadline = Date().addingTimeInterval(10)
    var accumulated = Data()
    while Date() < readyDeadline {
        let chunk = stdout.fileHandleForReading.availableData
        if !chunk.isEmpty { accumulated.append(chunk) }
        if let s = String(data: accumulated, encoding: .utf8), s.contains("READY") {
            readyReceived = true
            break
        }
        try await Task.sleep(nanoseconds: 20_000_000)
    }
    try #require(readyReceived)

    let clips = Locked<[ClipPayload]>([])
    let events = Locked<[DaemonEvent]>([])
    let peerNameBox = Locked<String?>(nil)
    // Tight per-link ping so the automatic keepalive surfaces a ping frame to
    // the fake peer within the test window; the fake peer pongs each ping so
    // the large deadFactor never trips the staleness dropper.
    let manager = LinkManager(
        config: LinkManager.LinkConfig(
            token: "interop-token", port: 28492, name: "swift-interop",
            appVersion: "0.0.0-test"),
        nodeID: UUID().uuidString.lowercased(),
        pingInterval: 0.5, pingDeadFactor: 20)
    await manager.setHandlers(
        onClip: { payload, peer in
            clips.set(clips.get() + [payload]); peerNameBox.set(peer)
        },
        emit: { events.set(events.get() + [$0]) })

    func waitUntil(_ timeout: Double, _ cond: @escaping () async -> Bool) async -> Bool {
        let deadline = monotonicNow() + timeout
        while monotonicNow() < deadline {
            if await cond() { return true }
            try? await Task.sleep(nanoseconds: 50_000_000)
        }
        return await cond()
    }

    // Link comes up with the Python peer after a routed dial.
    let outcome = await manager.tryConnect(
        to: .hostPort(host: "127.0.0.1", port: NWEndpoint.Port(rawValue: port)!),
        label: "127.0.0.1:\(port)")
    #expect(outcome == .routed)
    #expect(await waitUntil(5) { await manager.activeLinkCount() == 1 })

    // Python -> Swift text clip arrives, tagged with the peer's name.
    #expect(await waitUntil(5) {
        clips.get().contains {
            if case .text(let s) = $0 { return s == "hello-from-python" }
            return false
        }
    })
    #expect(peerNameBox.get() == "fake-peer")

    // Swift -> Python: text + image + file (+ automatic ping from the loop).
    _ = await manager.broadcast(.text("hello-from-swift"))
    _ = await manager.broadcast(.image(Data([0x89, 0x50, 0x4E, 0x47, 1, 2, 3])))
    _ = await manager.broadcast(.file(name: "노트.txt", data: Data("file-content".utf8)))

    #expect(await waitUntil(5) {
        guard let lines = try? String(contentsOf: outFile, encoding: .utf8) else {
            return false
        }
        return lines.contains("hello-from-swift")
            && lines.contains("\"kind\": \"file\"")
            && lines.contains("노트.txt")
            && lines.contains("\"kind\": \"image\"")
            && lines.contains("\"type\": \"ping\"")
    })

    // Swift -> a legacy (minor-0) Python peer: a two-file kind:"files" copy is
    // downgraded per-link to its first file on the wire (dropped == 1). The
    // second file never reaches the peer; the first does.
    let mf1 = (name: "노트-multi.txt", data: Data("files body one".utf8), relPath: String?.none)
    let mf2 = (name: "(E&S) plan.txt", data: Data("files body two".utf8), relPath: String?.none)
    let filesResult = await manager.broadcast(.files([mf1, mf2]))
    #expect(filesResult.maxDropped == 1)
    #expect(await waitUntil(5) {
        guard let lines = try? String(contentsOf: outFile, encoding: .utf8) else { return false }
        return lines.contains("노트-multi.txt")     // first file, sent as kind:"file"
            && !lines.contains("(E&S) plan.txt")    // second file dropped for the old peer
    })

    // The hello we sent must satisfy Python's field expectations.
    let outText = try String(contentsOf: outFile, encoding: .utf8)
    let helloLine = outText.split(separator: "\n").first { $0.contains("\"event\": \"hello\"") }
    let hello = try #require(helloLine)
    #expect(hello.contains("\"version\": 1"))
    #expect(hello.contains("\"protocol_major\": 1"))

    await manager.shutdown()
}

@Test func interopReceivesMultipleFilesFromFakePeer() async throws {
    let port: UInt16 = 28493
    let outFile = FileManager.default.temporaryDirectory
        .appendingPathComponent("fake-peer-\(UUID().uuidString).jsonl")

    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = [
        "python3", scriptsDir().appendingPathComponent("fake_peer.py").path,
        "--port", "\(port)", "--token", "interop-token",
        "--out", outFile.path, "--send-files",
    ]
    let stdout = Pipe()
    process.standardOutput = stdout
    try process.run()
    defer { if process.isRunning { process.terminate() } }

    var readyReceived = false
    let readyDeadline = Date().addingTimeInterval(10)
    var accumulated = Data()
    while Date() < readyDeadline {
        let chunk = stdout.fileHandleForReading.availableData
        if !chunk.isEmpty { accumulated.append(chunk) }
        if let s = String(data: accumulated, encoding: .utf8), s.contains("READY") {
            readyReceived = true
            break
        }
        try await Task.sleep(nanoseconds: 20_000_000)
    }
    try #require(readyReceived)

    let clips = Locked<[ClipPayload]>([])
    let events = Locked<[DaemonEvent]>([])
    let manager = LinkManager(
        config: LinkManager.LinkConfig(
            token: "interop-token", port: 28494, name: "swift-interop",
            appVersion: "0.0.0-test"),
        nodeID: UUID().uuidString.lowercased())
    await manager.setHandlers(
        onClip: { payload, _ in clips.set(clips.get() + [payload]) },
        emit: { events.set(events.get() + [$0]) })

    func waitUntil(_ timeout: Double, _ cond: @escaping () async -> Bool) async -> Bool {
        let deadline = monotonicNow() + timeout
        while monotonicNow() < deadline {
            if await cond() { return true }
            try? await Task.sleep(nanoseconds: 50_000_000)
        }
        return await cond()
    }

    let outcome = await manager.tryConnect(
        to: .hostPort(host: "127.0.0.1", port: NWEndpoint.Port(rawValue: port)!),
        label: "127.0.0.1:\(port)")
    #expect(outcome == .routed)
    #expect(await waitUntil(5) { await manager.activeLinkCount() == 1 })

    // Python -> Swift: the two-file batch surfaces with intact names, incl. the
    // parens/ampersand name the OLD whitelist would have mangled.
    #expect(await waitUntil(5) {
        clips.get().contains {
            if case .files(let fs) = $0 {
                return fs.count == 2
                    && fs.contains { $0.name == "노트.txt" && $0.data == Data("multi body one".utf8) }
                    && fs.contains { $0.name == "(E&S) plan.txt" && $0.data == Data("multi body two".utf8) }
            }
            return false
        }
    })
    // The received name is denylist-safe end-to-end: sanitize keeps it verbatim.
    #expect(sanitizeFilename("(E&S) plan.txt") == "(E&S) plan.txt")

    // Aggregate recomputed from the decoded bytes matches the CONTRACT formula.
    let expected = aggregateFilesHash([
        sha256Hex(Data("multi body one".utf8)), sha256Hex(Data("multi body two".utf8))])
    #expect(clips.get().contains {
        if case .files(let fs) = $0 {
            return aggregateFilesHash(fs.map { sha256Hex($0.data) }) == expected
        }
        return false
    })

    await manager.shutdown()
}

/// Spawn a wire-compatible fake peer that listens on `port`, handshakes, and
/// auto-sends one text clip; returns the process (for teardown) and its
/// record outfile. Waits for READY on stdout.
private func startFakePeer(port: UInt16, token: String) async throws -> (Process, URL) {
    let outFile = FileManager.default.temporaryDirectory
        .appendingPathComponent("fake-peer-\(UUID().uuidString).jsonl")
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
    process.arguments = [
        "python3", scriptsDir().appendingPathComponent("fake_peer.py").path,
        "--port", "\(port)", "--token", token, "--out", outFile.path,
    ]
    let stdout = Pipe()
    process.standardOutput = stdout
    try process.run()
    var ready = false
    let deadline = Date().addingTimeInterval(10)
    var accumulated = Data()
    while Date() < deadline {
        let chunk = stdout.fileHandleForReading.availableData
        if !chunk.isEmpty { accumulated.append(chunk) }
        if let s = String(data: accumulated, encoding: .utf8), s.contains("READY") {
            ready = true; break
        }
        try await Task.sleep(nanoseconds: 20_000_000)
    }
    guard ready else { throw WireConnectionError.closed }
    return (process, outFile)
}

/// Count of frames the fake peer RECEIVED from us (its "recv" event lines).
private func recvClipCount(_ url: URL) -> Int {
    guard let s = try? String(contentsOf: url, encoding: .utf8) else { return 0 }
    return s.split(separator: "\n").filter { $0.contains("\"event\": \"recv\"") }.count
}

private func fileContains(_ url: URL, _ needle: String) -> Bool {
    (try? String(contentsOf: url, encoding: .utf8))?.contains(needle) ?? false
}

@Test func meshLinksBothPeersAndNeverRelays() async throws {
    let portA: UInt16 = 28495
    let portB: UInt16 = 28496
    let (procA, outA) = try await startFakePeer(port: portA, token: "mesh-token")
    let (procB, outB) = try await startFakePeer(port: portB, token: "mesh-token")
    defer {
        if procA.isRunning { procA.terminate() }
        if procB.isRunning { procB.terminate() }
    }

    let clips = Locked<[(ClipPayload, String)]>([])
    let manager = LinkManager(
        config: LinkManager.LinkConfig(
            token: "mesh-token", port: 28497, name: "swift-mesh", appVersion: "0.0.0-test"),
        nodeID: UUID().uuidString.lowercased())
    await manager.setHandlers(
        onClip: { payload, peer in clips.set(clips.get() + [(payload, peer)]) },
        emit: { _ in })

    func waitUntil(_ timeout: Double, _ cond: @escaping () async -> Bool) async -> Bool {
        let deadline = monotonicNow() + timeout
        while monotonicNow() < deadline {
            if await cond() { return true }
            try? await Task.sleep(nanoseconds: 50_000_000)
        }
        return await cond()
    }

    // Dial BOTH peers: full mesh from this node's side.
    _ = await manager.tryConnect(
        to: .hostPort(host: "127.0.0.1", port: NWEndpoint.Port(rawValue: portA)!), label: "a")
    _ = await manager.tryConnect(
        to: .hostPort(host: "127.0.0.1", port: NWEndpoint.Port(rawValue: portB)!), label: "b")
    #expect(await waitUntil(5) { await manager.activeLinkCount() == 2 })

    // Both fake peers auto-send "hello-from-python"; the manager APPLIES both
    // locally (onClip records them). This proves both links carry traffic.
    #expect(await waitUntil(5) {
        clips.get().filter {
            if case .text(let s) = $0.0 { return s == "hello-from-python" }; return false
        }.count == 2
    })

    // Non-relay: the manager received A's (and B's) clip but has NO relay path,
    // so it must forward nothing to the other peer. Give it a bounded window,
    // then assert neither fake peer received ANY clip frame from us. (This is
    // also the echo-suppression-under-mesh guarantee: an applied clip is never
    // rebroadcast.)
    try await Task.sleep(nanoseconds: 1_000_000_000)
    #expect(recvClipCount(outA) == 0)
    #expect(recvClipCount(outB) == 0)

    // The mesh works the other way: one LOCAL clip reaches BOTH peers.
    _ = await manager.broadcast(.text("local-broadcast"))
    #expect(await waitUntil(5) {
        fileContains(outA, "local-broadcast") && fileContains(outB, "local-broadcast")
    })
    // That broadcast is the ONLY frame each peer received — still no relay of
    // the other peer's hello clip.
    #expect(await waitUntil(3) { recvClipCount(outA) == 1 })
    #expect(await waitUntil(3) { recvClipCount(outB) == 1 })

    await manager.shutdown()
}

File: forwindows/tests/AnyClipCore.Tests/InteropTests.cs
using System.Diagnostics;
using System.Runtime.CompilerServices;
using AnyClip.Core;
using Xunit;

namespace AnyClip.Core.Tests;

public class InteropTests
{
    private static string RepoRoot([CallerFilePath] string path = "") =>
        Path.GetFullPath(Path.Combine(Path.GetDirectoryName(path)!, "..", "..", ".."));

    // fake_peer.py keeps the out-file open for write for the whole session;
    // File.ReadAllText opens with FileShare.Read, a sharing violation on Windows
    // against that write handle. Do NOT "simplify" back to File.ReadAllText.
    private static string ReadShared(string path)
    {
        using var fs = new FileStream(path, FileMode.Open, FileAccess.Read,
            FileShare.ReadWrite | FileShare.Delete);
        using var r = new StreamReader(fs);
        return r.ReadToEnd();
    }

    private static ProcessStartInfo FakePeerPsi(int port, string outFile, bool sendFiles = false)
    {
        var psi = new ProcessStartInfo
        {
            FileName = "python3",
            ArgumentList =
            {
                Path.Combine(RepoRoot(), "formacOS", "Scripts", "fake_peer.py"),
                "--port", port.ToString(),
                "--token", "interop-token",
                "--out", outFile,
            },
            RedirectStandardOutput = true,
        };
        if (sendFiles) psi.ArgumentList.Add("--send-files");
        return psi;
    }

    private static async Task<bool> WaitUntil(Func<bool> cond, double seconds = 5)
    {
        var deadline = DateTime.UtcNow.AddSeconds(seconds);
        while (DateTime.UtcNow < deadline)
        { if (cond()) return true; await Task.Delay(50); }
        return cond();
    }

    [Fact]
    public async Task InteropWithPythonFakePeer()
    {
        int port = 28631;
        string outFile = Path.Combine(Path.GetTempPath(), $"fake-peer-{Guid.NewGuid()}.jsonl");
        using var proc = Process.Start(FakePeerPsi(port, outFile))!;
        try
        {
            var ready = await proc.StandardOutput.ReadLineAsync().WaitAsync(TimeSpan.FromSeconds(10));
            Assert.Equal("READY", ready);

            var clips = new List<(ClipPayload Payload, string Peer)>();
            var events = new List<DaemonEvent>();
            var manager = new LinkManager(
                new LinkConfig("interop-token", 28632, "csharp-interop", "0.0.0-test"),
                Guid.NewGuid().ToString().ToLowerInvariant());
            manager.OnClip = (p, peer) => { lock (clips) clips.Add((p, peer)); return Task.CompletedTask; };
            manager.Emit = e => { lock (events) events.Add(e); };

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
            await manager.TryConnectAsync("127.0.0.1", port, $"127.0.0.1:{port}", cts.Token);

            Assert.True(await WaitUntil(() => manager.ActiveLinkCount == 1));
            lock (events) Assert.Contains(events, e => e is LinkUp u && u.PeerName == "fake-peer");
            Assert.True(await WaitUntil(() =>
            { lock (clips) return clips.Any(c => c.Payload is TextClip t && t.Text == "hello-from-python"); }));

            await manager.BroadcastAsync(new TextClip("hello-from-csharp"));
            await manager.BroadcastAsync(new ImageClip(new byte[] { 0x89, 0x50, 0x4E, 0x47, 1 }));
            await manager.BroadcastAsync(new FileClip("노트.txt", "file-content"u8.ToArray()));

            Assert.True(await WaitUntil(() =>
            {
                if (!File.Exists(outFile)) return false;
                var lines = ReadShared(outFile);
                return lines.Contains("hello-from-csharp")
                    && lines.Contains("\"kind\": \"file\"")
                    && lines.Contains("노트.txt")
                    && lines.Contains("\"kind\": \"image\"");
            }));

            var outText = ReadShared(outFile);
            var helloLine = outText.Split('\n').FirstOrDefault(l => l.Contains("\"event\": \"hello\""));
            Assert.NotNull(helloLine);
            Assert.Contains("\"version\": 1", helloLine);
            Assert.Contains("\"protocol_major\": 1", helloLine);

            manager.Shutdown();
        }
        finally { if (!proc.HasExited) proc.Kill(); }
    }

    [Fact]
    public async Task InteropDowngradesFilesClipToOldPythonPeer()
    {
        int port = 28633;
        string outFile = Path.Combine(Path.GetTempPath(), $"fake-peer-{Guid.NewGuid()}.jsonl");
        using var proc = Process.Start(FakePeerPsi(port, outFile))!;
        try
        {
            var ready = await proc.StandardOutput.ReadLineAsync().WaitAsync(TimeSpan.FromSeconds(10));
            Assert.Equal("READY", ready);

            var manager = new LinkManager(
                new LinkConfig("interop-token", 28634, "csharp-interop", "0.0.0-test"),
                Guid.NewGuid().ToString().ToLowerInvariant());
            manager.OnClip = (_, _) => Task.CompletedTask;
            manager.Emit = _ => { };

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
            await manager.TryConnectAsync("127.0.0.1", port, $"127.0.0.1:{port}", cts.Token);
            Assert.True(await WaitUntil(() => manager.ActiveLinkCount == 1));

            // fake_peer advertises protocol_minor 0 -> the broadcast downgrades the
            // 2-file clip to the first file as a legacy "file", and reports 1 drop.
            var res = await manager.BroadcastAsync(new FilesClip(new List<(string, byte[])>
            {
                ("노트.txt", "multi body one"u8.ToArray()),
                ("(E&S) plan.txt", "multi body two"u8.ToArray()),
            }));
            Assert.Equal(1, res.OldPeerDrops);

            Assert.True(await WaitUntil(() =>
            {
                if (!File.Exists(outFile)) return false;
                var lines = ReadShared(outFile);
                return lines.Contains("\"kind\": \"file\"") && lines.Contains("노트.txt");
            }));
            // Never a multi-file "files" frame to a minor-0 peer.
            Assert.DoesNotContain("\"kind\": \"files\"", ReadShared(outFile));

            manager.Shutdown();
        }
        finally { if (!proc.HasExited) proc.Kill(); }
    }

    [Fact]
    public async Task InteropReceivesFilesClipFromPythonPeer()
    {
        int port = 28635;
        string outFile = Path.Combine(Path.GetTempPath(), $"fake-peer-{Guid.NewGuid()}.jsonl");
        using var proc = Process.Start(FakePeerPsi(port, outFile, sendFiles: true))!;
        try
        {
            var ready = await proc.StandardOutput.ReadLineAsync().WaitAsync(TimeSpan.FromSeconds(10));
            Assert.Equal("READY", ready);

            var clips = new List<(ClipPayload Payload, string Peer)>();
            var manager = new LinkManager(
                new LinkConfig("interop-token", 28636, "csharp-interop", "0.0.0-test"),
                Guid.NewGuid().ToString().ToLowerInvariant());
            manager.OnClip = (p, peer) => { lock (clips) clips.Add((p, peer)); return Task.CompletedTask; };
            manager.Emit = _ => { };

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
            await manager.TryConnectAsync("127.0.0.1", port, $"127.0.0.1:{port}", cts.Token);
            Assert.True(await WaitUntil(() => manager.ActiveLinkCount == 1));
            Assert.True(await WaitUntil(() =>
            { lock (clips) return clips.Any(c => c.Payload is FilesClip f && f.Files.Count == 2); }));

            FilesClip got;
            lock (clips) got = clips.Select(c => c.Payload).OfType<FilesClip>().First(f => f.Files.Count == 2);
            Assert.Equal("노트.txt", got.Files[0].Name);
            Assert.Equal("multi body one", System.Text.Encoding.UTF8.GetString(got.Files[0].Data));
            Assert.Equal("(E&S) plan.txt", got.Files[1].Name);
            Assert.Equal("multi body two", System.Text.Encoding.UTF8.GetString(got.Files[1].Data));

            manager.Shutdown();
        }
        finally { if (!proc.HasExited) proc.Kill(); }
    }

    [Fact]
    public async Task InteropTwoPeersReceiveBroadcastAndNoRelay()
    {
        int portA = 28637, portB = 28638;
        string outA = Path.Combine(Path.GetTempPath(), $"fake-peer-A-{Guid.NewGuid()}.jsonl");
        string outB = Path.Combine(Path.GetTempPath(), $"fake-peer-B-{Guid.NewGuid()}.jsonl");
        using var procA = Process.Start(FakePeerPsi(portA, outA))!;
        using var procB = Process.Start(FakePeerPsi(portB, outB))!;
        try
        {
            Assert.Equal("READY", await procA.StandardOutput.ReadLineAsync().WaitAsync(TimeSpan.FromSeconds(10)));
            Assert.Equal("READY", await procB.StandardOutput.ReadLineAsync().WaitAsync(TimeSpan.FromSeconds(10)));

            var clips = new List<(ClipPayload Payload, string Peer)>();
            var manager = new LinkManager(
                new LinkConfig("interop-token", 28639, "csharp-interop", "0.0.0-test"),
                Guid.NewGuid().ToString().ToLowerInvariant());
            manager.OnClip = (p, peer) => { lock (clips) clips.Add((p, peer)); return Task.CompletedTask; };
            manager.Emit = _ => { };

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
            await manager.TryConnectAsync("127.0.0.1", portA, $"127.0.0.1:{portA}", cts.Token);
            await manager.TryConnectAsync("127.0.0.1", portB, $"127.0.0.1:{portB}", cts.Token);
            Assert.True(await WaitUntil(() => manager.ActiveLinkCount == 2));

            // Both peers pushed their own "hello-from-python" clip; both applied locally.
            Assert.True(await WaitUntil(() =>
            { lock (clips) return clips.Count(c => c.Payload is TextClip t && t.Text == "hello-from-python") >= 2; }));

            // One local clip broadcasts to BOTH peers.
            await manager.BroadcastAsync(new TextClip("mesh-broadcast"));
            Assert.True(await WaitUntil(() => File.Exists(outA) && ReadShared(outA).Contains("mesh-broadcast")));
            Assert.True(await WaitUntil(() => File.Exists(outB) && ReadShared(outB).Contains("mesh-broadcast")));

            // Non-relay: peer A's clip was applied locally, NEVER forwarded to B.
            // Drain B for a bounded interval; every text clip B received is our
            // broadcast, never a relayed "hello-from-python" (this doubles as the
            // echo-suppression-under-mesh check).
            await Task.Delay(1000);
            var recvTextFrames = ReadShared(outB).Split('\n')
                .Where(l => l.Contains("\"event\": \"recv\"") && l.Contains("\"kind\": \"text\""))
                .ToList();
            Assert.NotEmpty(recvTextFrames);
            Assert.All(recvTextFrames, l => Assert.DoesNotContain("hello-from-python", l));
            Assert.Contains(recvTextFrames, l => l.Contains("mesh-broadcast"));

            manager.Shutdown();
        }
        finally
        {
            if (!procA.HasExited) procA.Kill();
            if (!procB.HasExited) procB.Kill();
        }
    }

    [Fact]
    public async Task InteropFolderOnlyClipSendsNothingToTheMinorZeroPythonPeer()
    {
        int port = 28641;
        string outFile = Path.Combine(Path.GetTempPath(), $"fake-peer-{Guid.NewGuid()}.jsonl");
        using var proc = Process.Start(FakePeerPsi(port, outFile))!;
        try
        {
            var ready = await proc.StandardOutput.ReadLineAsync().WaitAsync(TimeSpan.FromSeconds(10));
            Assert.Equal("READY", ready);

            var manager = new LinkManager(
                new LinkConfig("interop-token", 28642, "csharp-interop", "0.0.0-test"),
                Guid.NewGuid().ToString().ToLowerInvariant());
            manager.OnClip = (_, _) => Task.CompletedTask;
            manager.Emit = _ => { };

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
            await manager.TryConnectAsync("127.0.0.1", port, $"127.0.0.1:{port}", cts.Token);
            Assert.True(await WaitUntil(() => manager.ActiveLinkCount == 1));

            // Every entry came from a copied folder, and protocol 1.0's single
            // kind:"file" frame has nowhere to put a path -> nothing is sent to
            // this peer at all.
            var res = await manager.BroadcastAsync(new FilesClip(new List<FileEntry>
            {
                new("a.txt", "one"u8.ToArray(), "docs/a.txt"),
                new("b.txt", "two"u8.ToArray(), "docs/sub/b.txt"),
            }));
            Assert.Empty(res.Delivered);
            Assert.Equal(0, res.OldPeerDrops);
            // Skipping is NOT dropping the link: the peer stays connected and a
            // following ordinary clip still reaches it.
            Assert.Equal(1, manager.ActiveLinkCount);

            // Sentinel: waiting on a clip that DOES arrive is what turns
            // "nothing was written" into a real assertion instead of a race
            // against a still-empty file.
            await manager.BroadcastAsync(new TextClip("after-folder"));
            Assert.True(await WaitUntil(() =>
                File.Exists(outFile) && ReadShared(outFile).Contains("after-folder")));
            var seen = ReadShared(outFile);
            Assert.DoesNotContain("\"kind\": \"files\"", seen);
            Assert.DoesNotContain("a.txt", seen);

            manager.Shutdown();
        }
        finally { if (!proc.HasExited) proc.Kill(); }
    }
}
